src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -26,23 +36,67 @@ public void setProductDao(ProductDao productDao) {
     }
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        String requestPath = request.getPathInfo().substring(1);
+        String requestPath = getRequestPath(request);
         if (requestPath.isEmpty()) {
             response.sendError(HttpServletResponse.SC_NOT_FOUND);
         }
-        String[] pathParameters = requestPath.split("/");
-        Long productId = Optional.of(pathParameters[0]).map(Long::valueOf).orElse(null);
-        request.setAttribute("product", productDao.getProduct(productId));
-        if (pathParameters.length == 1) {
-            request.getRequestDispatcher("/WEB-INF/pages/productDetails.jsp").forward(request, response);
-        } else if ("priceHistory".equals(pathParameters[1])) {
-            request.getRequestDispatcher("/WEB-INF/pages/productPriceHistory.jsp").forward(request, response);
+        String[] requestPathParameters = splitRequestPathParameters(requestPath);
+        Long productId = Optional.of(requestPathParameters[0]).map(Long::valueOf).orElse(null);
+        setProductAttribute(request, productDao.getProduct(productId));
+        if (requestPathParameters.length == 1) {
+            request.getRequestDispatcher(PRODUCT_DETAILS_JSP).forward(request, response);
+        } else if ("priceHistory".equals(requestPathParameters[1])) {
+            request.getRequestDispatcher(PRODUCT_PRICE_HISTORY_JSP).forward(request, response);
         } else {
             response.sendError(HttpServletResponse.SC_NOT_FOUND);
         }
     }
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        String[] requestPathParameters = splitRequestPathParameters(getRequestPath(request));
+        Long productId = Optional.of(requestPathParameters[0]).map(Long::valueOf).orElse(null);
+        NumberFormat numberFormat = NumberFormat.getInstance(request.getLocale());
+        String quantityString = request.getParameter("quantity");
+        try {
+            cartService.add(productId, numberFormat.parse(quantityString).intValue());
+            String successMessage = "Product was added to cart";
+            response.sendRedirect(String.format("%s/products/%d?message=%s",
+                    request.getContextPath(), productId, successMessage));
+            return;
+        } catch (ParseException parseException) {
+            setAddToCartErrorAttribute(request, "Not a number!");
+        } catch (OutOfStockException outOfStockException) {
+            setAddToCartErrorAttribute(request,
+                    String.format("Not enough items in stock! Available: %d",
+                            outOfStockException.getAvailableQuantity()
+                    )
+            );
+        }
+        setProductAttribute(request, productDao.getProduct(productId));
+        request.setAttribute("cart", cartService.getCart());
```

Comment:
it's better to reuse the doGet() method as it already implements the necessary logic to display the productDetails page
In this regard, the price history logic should be moved to a separate servlet so as not to mix the functionality